src/history.js

```diff
@@ -187,13 +187,13 @@ class History extends Emitter {
     if (this.size <= 0) {
       this.begin()
       return
-    } else if (!next) {
+    } else if (action === this.head) {
       next = this.head = parent
     } else if (action === this.root) {
       this.root = next
     }
 
-    if (!action.disabled) {
+    if (!action.disabled && next) {
       this.reconcile(next)
     }
```

Comment:
i feel like this probably needs a reconsideration as well, my guess is that we should only call `reconcile` with `this.head` if the head has changed, but wanted to run a gut check on that intuition before pursuing.